pkg/apis/policy/types.go

```diff
@@ -68,3 +68,53 @@ type PodDisruptionBudgetList struct {
 	unversioned.ListMeta `json:"metadata,omitempty"`
 	Items                []PodDisruptionBudget `json:"items"`
 }
+
+// NetworkPolicy is an object to define network policies that can be applied to pods.
+type NetworkPolicy struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object metadata; More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior for this NetworkPolicy.
+	Spec NetworkPolicySpec `json:"spec,omitempty"`
+}
+
+type NetworkPolicySpec struct {
+	// Selects the pods to which this NetworkPolicy object applies.
+	PodSelector unversioned.LabelSelector `json:"podSelector"`
+
+	// List of ingress rules to be applied to the selected pods.
+	Ingress []NetworkPolicyIngressRule `json:"ingress,omitempty"`
+}
+
+// This NetworkPolicyIngressRule matches traffic if and only if the traffic matches both Ports AND From.
+type NetworkPolicyIngressRule struct {
+	// List of ports which should be made accessible on the pods selected by PodSelector.
```

Comment:
The proposal doc had lots of good comments - what happened?  Specifically "if not provided, match all" and "if empty match none", although we can't really implement that.  Go conflates empty and non-specified with "omitempty".  I think we can only say that unspecified or empty means no ports are allowed.

We can make it a pointer to slice, if we want to keep the semantic.  https://play.golang.org/p/-yCdDclY7_